### PR TITLE
ECO-1067 - Make Jenkins Pipeline support "Run Sauce Labs Test Publisher"

### PR DESCRIPTION
**Pull request by halkeye**
* Mostly changes to switch from AbstractBuild to Run where appropriate
* BuildAction will now store credentials id so it doesn't need to look
  it up in the build wrapper so it can display results
* Project action will now be added on the fly by the build action
* Broke support zip usefulness a bit, will have to rething that.

### CODE REVIEW

**Comment by halkeye** [comment]
Kinda WIP, @moizjv

File: src/main/java/hudson/plugins/sauce_ondemand/SauceEnvironmentUtil.java [hunk 1]
```diff
@@ -213,7 +210,7 @@ public static SauceOnDemandBuildWrapper getBuildWrapper(AbstractProject<?, ?> pr
      * @return String representing the Jenkins build
      */
     @Nonnull
-    public static String getBuildName(AbstractBuild<?, ?> build) {
+    public static String getBuildName(Run<?, ?> build) {
```
Comment (moizjv) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceEnvironmentUtil.java]: is it possible for putting map types for example `Run<String, String>` ?

File: src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java [hunk 2]
```diff
@@ -25,42 +33,39 @@
...
-     */
-    private static final String JOB_DETAILS_URL = "http://saucelabs.com/rest/v1/%1$s/build/%2$s/jobs?full=true";
-
-    /**
      * Regex pattern that is used to identify Sauce job ids which have been run as part of a Jenkins build.
      */
     public static final Pattern SESSION_ID_PATTERN = Pattern.compile("SauceOnDemandSessionID=([0-9a-fA-F]+)(?:.job-name=(.*))?");
 
-    private AbstractBuild<?, ?> build;
+    private Run build;
```
Comment (moizjv) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java]: why not call it `Run run`?

File: src/main/java/hudson/plugins/sauce_ondemand/SauceEnvironmentUtil.java (hunk 1 above)
Comment (halkeye) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceEnvironmentUtil.java]: Thats not a map, its a generic type. It pretty much means any sub types.

File: src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java (hunk 2 above)
Comment (halkeye) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java]: I'd have to write migration code, leave both variables in place for a while. Jenkins de-serializes xml files. You can take a look in the work/ directory

File: src/main/java/hudson/plugins/sauce_ondemand/SauceEnvironmentUtil.java (hunk 1 above)
Comment (moizjv) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceEnvironmentUtil.java]: Yeah that is interesting, I have always seen it for maps `Map<K,V>`, you are right ! :).

File: src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java (hunk 2 above)
Comment (halkeye) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java]: `./work/jobs/TestSimple-Java-TestNG/builds/1/build.xml` shows:
```
    <hudson.plugins.sauce__ondemand.SauceOnDemandBuildAction plugin="sauce-ondemand@1.148-SNAPSHOT">
      <build class="build" reference="../../.."/>
      <jobInformation>
        <hudson.plugins.sauce__ondemand.JenkinsJobInformation>
```

The build variable is mapped to higher level. Whether it be a freestyle build, maven build, pipeline build, etc.

If i change the variable name, the de-serialization will break unless i do migration code.

AbstractBuild and Run are very similar. Its higher up in the chain and will work with the pipeline classes. Its still essentially the build object.

Comment (moizjv) [inline src/main/java/hudson/plugins/sauce_ondemand/SauceOnDemandBuildAction.java]: yup, I agree if variable name is tightly coupled does not make sense to break it.

**Comment by halkeye** [comment]
@moizjv did you finish the review?

**Comment by moizjv** [comment]
LGTM

**Pull request merged**